core-bundle/src/Filesystem/Dbafs.php

```diff
@@ -0,0 +1,861 @@
+                $this->table,
+                ['path' => $this->convertToDatabasePath($pathToDelete)]
+            );
+        }
+
+        $this->connection->commit();
+    }
+
+    /**
+     * Loads paths from the database that should be considered when synchronizing.
+     * his includes all parent directories and - in case of directories - all
+     * resources that reside in it.
+     *
+     * This method also builds lookup tables for hashes and UUIDs of the entire
+     * table.
+     *
+     * @param array<int, string> $searchPaths       non-empty list of search paths
+     * @param array<int, string> $parentDirectories parent directories to consider
+     *
+     * @return array<int, array<int|string, string>>
+     * @phpstan-return array{0: array<int, string>, 1: array<string, string>, 2: array<string, string>}
+     */
+    private function getDatabaseEntries(array $searchPaths, array $parentDirectories): array
+    {
+        $dbPaths = [];
+        $allHashesByPath = [];
+        $allUuidsByPath = [];
+
+        $items = $this->connection->fetchAllNumeric(
+            sprintf(
+                "SELECT path, uuid, hash, IF(type='folder', 1, 0) AS is_dir FROM %s",
+                $this->connection->quoteIdentifier($this->table)
+            )
+        );
+
+        $fullScope = '' === $searchPaths[0];
+
+        foreach ($items as [$path, $uuid, $hash, $isDir]) {
+            $path = $this->convertToFilesystemPath($path);
+
+            // Include a path if it is either inside the search paths or is a
+            // parent directory of it.
+            $include = $fullScope || $this->inPath($path, $searchPaths) || ($isDir && \in_array($path, $parentDirectories, true));
+
+            // Make directories distinguishable by appending a slash
+            if ($isDir) {
+                $path .= '/';
+            }
+
+            if ($include) {
+                $dbPaths[] = $path;
+            }
+
+            $allHashesByPath[$path] = $hash;
+            $allUuidsByPath[$path] = $uuid;
+        }
+
+        return [$dbPaths, $allHashesByPath, $allUuidsByPath];
+    }
+
+    /**
+     * Traverses the filesystem and returns file and directory paths that can
+     * be synchronized. For files, a hash of the content is returned alongside,
+     * for directories null, for none-existing resources false.
+     *
+     * Items will always be listed before the directories they reside in (most
+     * specific path first).
+     *
+     * @param array<int, string> $searchPaths       non-empty list of search paths
+     * @param array<int, string> $parentDirectories parent directories to consider
+     *
+     * @return \Generator<string, (string|false|null)>
+     */
+    private function getFilesystemPaths(FilesystemAdapter $filesystem, array $searchPaths, array $parentDirectories): \Generator
+    {
+        $getHash = fn (string $path): string => hash($this->hashAlgorithm, $filesystem->read($path));
```

Comment:
```suggestion
        $getHash = function (string $path) use ($filesystem, $path): string {
            $ctx = hash_init($this->hashAlgorithm);
            hash_update_stream($ctx, $filesystem->readStream($path));
            return hash_final($ctx);
        };
```

Not sure if it works but that would be a lot more memory efficient ;)

Reply:
Works! 

(and is already implemented in 7b97f8cf41d13bfcd8f14290b91a2f6bde97d3e5)